### PR TITLE
Soil moisture analysis

### PR DESCRIPTION
**Pull request by prske**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,10 +43,114 @@ Examples:
 <h3>POST</h3>
 /api/v1/eto/
```
Comment (fedjo) [inline README.md:44]: This method documentation is not clear. POST with what kind of input data?

File: README.md [hunk 2]
```diff
@@ -43,10 +43,114 @@ Examples:
...
+```
+
+```json
+{
+  "status_code":201, 
+  "detail": "Successfully deleted"
+}
+```
+
+<h3>POST</h3>
```
Comment (fedjo) [inline README.md:79]: Please also define the format of the input dataset

File: README.md [hunk 3]
```diff
@@ -43,10 +43,114 @@ Examples:
...
+  ]
+}
+```
+
+
 <h3> Example usage </h3>
 
 In order to use the aforementioned API, you need to fill out the input parameters corresponding to the requested ETO calculation.
 
+Use POST /api/v1/dataset/{dataset_id} to upload your data to the database.
```
Comment (fedjo) [inline README.md:149]: maybe `POST /api/v1/dataset/` ?

File: app/api/api_v1/endpoints/dataset.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+from schemas import Dataset as DatasetScheme
+from db import SessionLocal, engine
+from models import Dataset
+from crud import get_datasets, add_dataset, delete_datasets
+from utils import (min_max_date, detect_irrigation_events, count_precipitation_events,
+                   count_high_dose_irrigation_events, get_high_dose_irrigation_events_dates, calculate_field_capacity,
+                   calculate_stress_level, get_stress_count, get_stress_dates,
+                   no_of_saturation_days, get_saturation_dates)
+
+Dataset.metadata.create_all(bind=engine)
```
Comment (fedjo) [inline app/api/api_v1/endpoints/dataset.py:15]: I don't think it as a good practice to create the db tables within the route files. That way every time you restart the application you try to create the tables in the db

File: app/api/api_v1/endpoints/dataset.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+def get_db():
```
Comment (fedjo) [inline app/api/api_v1/endpoints/dataset.py:17]: For code maintanance reasons, I suggest to add this method in a separate `db.py` file or in `database.py`

File: app/schemas/dataset_scheme.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+from pydantic import BaseModel, ConfigDict
+from typing import List, Tuple
+from datetime import datetime
+
+class Dataset(BaseModel):
+    model_config = ConfigDict(from_attributes=True)
+
+    dataset_id: int
```
Comment (fedjo) [inline app/schemas/dataset_scheme.py:8]: what is this id? where does it come from? is it an identifier?

File: app/crud/dataset_operations.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+from sqlalchemy.orm import Session
+
+from models import Dataset as DM
+from schemas import Dataset as DS
+
+def add_dataset(db: Session, dataset: DS):
+    db_dataset = DM(dataset_id = dataset.dataset_id,
```
Comment (fedjo) [inline app/crud/dataset_operations.py:7]: maybe you can do sth like

```
from fastapi.encoders import jsonable_encoder
dataset_json = jsonable_encoder(dataset)
db_dataset = DM(**dataset_json)
```

File: app/models/dataset_model.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+from sqlalchemy import Column, Integer, String, Float
+
+from db import Base
+
+class Dataset(Base):
+    __tablename__ = "dataset"
+
+    id = Column(Integer, primary_key=True)
+    dataset_id = Column(Integer)
+    date = Column(String)
```
Comment (fedjo) [inline app/models/dataset_model.py:10]: does saving date as string allows date queries?

File: requirements.txt [hunk 9]
```diff
@@ -1,7 +1,7 @@
 fastapi==0.111.0
 alembic==1.13.1 # DB migrations
 requests==2.32.3 # Readable HTTP
-psycopg2==2.9.9 # PSQL driver
+psycopg2-binary==2.9.9 # PSQL driver
 pydantic-settings==2.2.1 # Pydantic settings options [donated to the python-org, not part of the main package anymore]
 PyJWT==2.8.0 # Instead of jose (has a CVE)
```
Comment (fedjo) [inline requirements.txt:6]: I think this dependency is not needed, as long as, the app does not perform any authentication so far

File: requirements.txt [hunk 10]
```diff
@@ -1,7 +1,7 @@
 fastapi==0.111.0
 alembic==1.13.1 # DB migrations
```
Comment (fedjo) [inline requirements.txt:2]: is it needed?

File: app/api/api_v1/endpoints/dataset.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+router = APIRouter()
+
+@router.post("/")
+def load_dataset(dataset: DatasetScheme, db: Session = Depends(get_db)) -> DatasetScheme:
+    return add_dataset(db, dataset)
+
+
+@router.get("/{dataset_id}")
+async def get_dataset(dataset_id: int, db: Session = Depends(get_db)) -> list[DatasetScheme]:
+    db_dataset = get_datasets(db, dataset_id)
```
Comment (fedjo) [inline app/api/api_v1/endpoints/dataset.py:33]: i guess this is a list `db_datasets`

File: app/api/api_v1/endpoints/dataset.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+router = APIRouter()
+
+@router.post("/")
+def load_dataset(dataset: DatasetScheme, db: Session = Depends(get_db)) -> DatasetScheme:
+    return add_dataset(db, dataset)
+
+
+@router.get("/{dataset_id}")
+async def get_dataset(dataset_id: int, db: Session = Depends(get_db)) -> list[DatasetScheme]:
```
Comment (fedjo) [inline app/api/api_v1/endpoints/dataset.py:32]: if this method return a list maybe change to `get_datasets`

File: app/api/api_v1/endpoints/dataset.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    except:
+        raise HTTPException(status_code=400, detail="Could not delete dataset")
+
+    if deleted == 0:
+        raise HTTPException(status_code=400, detail="No dataset with given id")
+    return {"status_code":201, "detail": "Successfully deleted"}
+
+@router.get("/{dataset_id}/analysis")
+def analyse_soil_moisture(dataset_id: int, db: Session = Depends(get_db)) -> DatasetAnalysis:
+    dataset: list[DatasetScheme] = get_datasets(db, dataset_id)
```
Comment (fedjo) [inline app/api/api_v1/endpoints/dataset.py:51]: Is this a list of a single dataset, I got confused? :-Z

File: app/db/database.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
+from sqlalchemy import create_engine
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+
+SQLALCHEMY_DATABASE_URL = "sqlite:///./sql_app.db"
```
Comment (fedjo) [inline app/db/database.py:5]: maybe is a better idea to pass this value from an environment variable

File: app/db/database.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+
+engine = create_engine(SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False})
```
Comment (fedjo) [inline app/db/database.py:7]: what is the `check_same_thread` do?

File: app/utils/dataset_utils.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
+from schemas import Dataset as DatasetScheme
+
+def min_max_date(dataset: list[DatasetScheme]):
```
Comment (fedjo) [inline app/utils/dataset_utils.py:3]: Add return types for more clear code

File: app/utils/dataset_utils.py [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+    min_date = min(dataset, key=lambda d: d.date)
+    max_date = max(dataset, key=lambda d: d.date)
+    return [min_date, max_date]
+
+
+def detect_irrigation_events(dataset: list[DatasetScheme]):
+    const_threshold = 0.01
```
Comment (fedjo) [inline app/utils/dataset_utils.py:10]: Where do these static values come from? Maybe import them from a config/env file in order to have the ability to change them in the future

File: app/utils/dataset_utils.py [hunk 18]
```diff
@@ -0,0 +1,236 @@
+from schemas import Dataset as DatasetScheme
+
+def min_max_date(dataset: list[DatasetScheme]):
+    min_date = min(dataset, key=lambda d: d.date)
+    max_date = max(dataset, key=lambda d: d.date)
+    return [min_date, max_date]
+
+
+def detect_irrigation_events(dataset: list[DatasetScheme]):
```
Comment (fedjo) [inline app/utils/dataset_utils.py:9]: I would also suggest to add as a comment in all these method a short description of the logic you are following

File: README.md (hunk 1 above)
Comment (prske) [reply README.md:44]: This is not part of this PR. It is going to be implemented by Stefan, like it was mentioned on our last sync call

File: app/db/database.py (hunk 15 above)
Comment (prske) [reply app/db/database.py:7]: https://fastapi.tiangolo.com/tutorial/sql-databases/#create-models:~:text=on%20the%20database).-,Create%20an%20Engine,-%C2%B6

File: requirements.txt (hunk 9 above)
Comment (prske) [reply requirements.txt:6]: This dependency is for SQL not authentication
https://pypi.org/project/psycopg2-binary/

File: requirements.txt (hunk 10 above)
Comment (prske) [reply requirements.txt:2]: yes, if not now in the future definitely

File: app/api/api_v1/endpoints/dataset.py (hunk 5 above)
Comment (prske) [reply app/api/api_v1/endpoints/dataset.py]: solved with 61c843df7467a3e547638d772de8a65fabb79688 commit

File: app/models/dataset_model.py (hunk 8 above)
Comment (prske) [reply app/models/dataset_model.py:22]: As far as I know, we do not need to query dates

File: app/utils/dataset_utils.py (hunk 16 above)
Comment (prske) [reply app/utils/dataset_utils.py]: Solved with 61c843df7467a3e547638d772de8a65fabb79688 commit

File: app/utils/dataset_utils.py (hunk 17 above)
Comment (prske) [reply app/utils/dataset_utils.py]: Solved with 61c843df7467a3e547638d772de8a65fabb79688 commit

File: app/db/database.py (hunk 14 above)
Comment (prske) [reply app/db/database.py]: Solved with 670527f00942c9c7b2d875fe9a817a32c353848b commit

File: app/api/api_v1/endpoints/dataset.py (hunk 4 above)
Comment (prske) [reply app/api/api_v1/endpoints/dataset.py]: Discussed on Friday

File: app/schemas/dataset_scheme.py (hunk 6 above)
Comment (prske) [reply app/schemas/dataset_scheme.py:8]: Discussed on Friday

File: app/crud/dataset_operations.py (hunk 7 above)
Comment (prske) [reply app/crud/dataset_operations.py:7]: Discussed on Friday

File: app/api/api_v1/endpoints/dataset.py (hunk 11 above)
Comment (prske) [reply app/api/api_v1/endpoints/dataset.py:27]: Discussed on Friday

File: app/api/api_v1/endpoints/dataset.py (hunk 12 above)
Comment (prske) [reply app/api/api_v1/endpoints/dataset.py:26]: Discussed on Friday

File: app/api/api_v1/endpoints/dataset.py (hunk 13 above)
Comment (prske) [reply app/api/api_v1/endpoints/dataset.py:48]: Discussed on Friday

File: app/utils/dataset_utils.py (hunk 18 above)
Comment (prske) [reply app/utils/dataset_utils.py]: Solved with 10bfe026d409b6764382f8d87866d8c9f4875fc5 commit

File: app/utils/dataset_utils.py [hunk 19]
```diff
@@ -1,14 +1,27 @@
...
     max_date = max(dataset, key=lambda d: d.date)
     return [min_date, max_date]
 
 
-def detect_irrigation_events(dataset: list[DatasetScheme]):
-    const_threshold = 0.01
-    increase_threshold = 0.05
+"""
+With two thresholds we detect irrigation events.
+First thresholds checks the moisture white the second one checks if there was increase.
```
Comment (fedjo) [inline app/utils/dataset_utils.py:18]: typo "white" -> "while"

Comment (fedjo) [reply app/utils/dataset_utils.py:18]: Maybe also describe what is the input (also for all methods)


Eg
```
Read a dataset which contains infro regarding temperature ... and tries to detect irrigation events
...
etc
 ```

File: app/utils/dataset_utils.py
Comment (fedjo) [inline app/utils/dataset_utils.py:1]: General comments

1. Comment sections should give a bit more info regarding input params, algorith, output param.
```
# Method fun() takes as input integer and a string and tries to identify ... 
def fun(a: int, b: str) -> int

or 

# Method fun() identifies ... in a string
#
# a: an integer...
# b: a string ...
# returns: the position...
def fun(a:int, b:string) -> int
```

2. I noticed that methods with similar usage have use same code. Maybe that's a good reason to export this code to a separate function eg `detect_irrigation` and `count_precipitation_events`. Also pay attention to the method names. If the two above methods do similiar stuff changing _irrigation_events_ to _precipitation_events_ creates confusion

File: app/core/co.py
Comment (fedjo) [inline app/core/co.py:1]: is this file needed?

File: app/api/api_v1/endpoints/dataset.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+    if not dataset:
+        raise HTTPException(status_code=404, detail="Dataset not found")
+
+    result = DatasetAnalysis
+
+    result.dataset_id = dataset_id
+    result.time_period = min_max_date(dataset)
+    result.irrigation_events_detected = detect_irrigation_events(dataset, settings)
```
Comment (fedjo) [inline app/api/api_v1/endpoints/dataset.py:57]: Hmmm this is a matter of code taste comment.

Personally, I would not choose to pass `settings` object as an argument to all these methods. Even though the solution has correct implementation,  semantically the method `detect_irrigation_events` should not take as an input parameter the application config file (`settings` obejct).

Instead I would keep the method signature as is `def detect_irrigation_events(dataset)` and just read the settings from the config file.

In the file `utils/dataset_utils`
```
from core import get_settings

settings = get_settings()


def detect_irrigation_events(dataset: list[DatasetScheme]):
    const_threshold = settings.CONST_THRESHOLD
    increase_threshold = settings.INCREASE_THRESHOLD
...
```

Both implementation work fine. I leave up to you to decide!

File: app/core/co.py
Comment (prske) [reply app/core/co.py:1]: Had some idea with that... Deleted

File: app/utils/dataset_utils.py
Comment (prske) [reply app/utils/dataset_utils.py:1]: Yes, I agree with you. 
I was also confused when I got documents with the specification, and for me, more or less everything is the same, as it was implemented.

I also agree with you that we should somehow move same logic into separate function (or maybe cashing).
This could be solved if we have .csv file into database instead of this approach.
Then we would have less functions and also isolate some of them.

Let me know what you think. Also comment it on Issue #7

File: app/api/api_v1/endpoints/dataset.py (hunk 20 above)
Comment (prske) [reply app/api/api_v1/endpoints/dataset.py]: Now that I got second look, it seems that this approach is better.
If we think long distance, at some point, setting could be much bigger and that can cause *problems*.

Thanks!

Comment (fedjo) [reply app/api/api_v1/endpoints/dataset.py]: Great! Comment resolved

File: app/utils/dataset_utils.py
Comment (fedjo) [reply app/utils/dataset_utils.py:1]: Ok, I understand your point. What I propose then is to merge this PR and continue the discussion in the issue.

Nevertheless, dataset or method refactoring will be done in another PR.

If you agree then you can resolve and merge the PR

**Review by fedjo — DISMISSED** [review]

**Comment by prske** [comment]
Hi @fedjo can you just approve this again? I have made a commit in order to avoid potential merge conflicts in README.md

**Comment by fedjo** [comment]
> Hi @fedjo can you just approve this again? I have made a commit in order to avoid potential merge conflicts in README.md

Nice, but I still see some conflicts. I think you need to rebase this branch (and all other branches) on top of main branch and resolve the conflicts.

If you are not familiar with rebase we can always have a pair-programming call :-)

**Comment by prske** [comment]
It should be OK now

**Review by fedjo — APPROVED** [review]
